### ISSUE TITLE
Algolia Search returns objects but they aren't visible - Uncaught (in promise) TypeError: e.hierarchy is undefined

### ISSUE THREAD

**Issue by ottozastrow** [facebook/docusaurus#7921]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

building a site to help people working with medical device regulations.

When inspecting network packets on the site (regardless of netlify or localhost) I see search results in the algolia response but the search widget doesn't show them.

When search is used the console throws this error
`Uncaught (in promise) TypeError: e.hierarchy is undefined`

This bug occurs even with a fresh project 
`npx create-docusaurus@latest standlet_fresh2 classic` (and the typescript version)

(I disabled contextualSearch, set APIkey, AppId, indexName) 

Here my index config from algolia
perhaps its related to searchableAttributes? I tried adding hierachry.lvl0 .. lvl6, but that didn't work (also seemed random)

``{
  "settings": {
    "minWordSizefor1Typo": 4,
    "minWordSizefor2Typos": 8,
    "hitsPerPage": 20,
    "maxValuesPerFacet": 100,
    "searchableAttributes": [
      "unordered(content)",
      "unordered(section)",
      "unordered(title)",
      "unordered(permalink)"
    ],
    "numericAttributesToIndex": null,
    "attributesToRetrieve": null,
    "allowTyposOnNumericTokens": false,
    "unretrievableAttributes": null,
    "optionalWords": null,
    "queryLanguages": [],
    "attributesForFaceting": [
      "content",
      "section",
      "title"
    ],
    "attributesToSnippet": null,
    "attributesToHighlight": null,
    "paginationLimitedTo": 1000,
    "attributeForDistinct": "permalink",
    "exactOnSingleWordQuery": "attribute",
    "ranking": [
      "typo",
      "geo",
      "words",
      "filters",
      "proximity",
      "attribute",
      "exact",
      "custom"
    ],
    "customRanking": null,
    "separatorsToIndex": "",
    "removeWordsIfNoResults": "none",
    "queryType": "prefixLast",
    "highlightPreTag": "<em>",
    "highlightPostTag": "</em>",
    "snippetEllipsisText": "",
    "alternativesAsExact": [
      "ignorePlurals",
      "singleWordSynonym"
    ],
    "indexLanguages": [],
    "renderingContent": {
      "facetOrdering": {
        "facets": {
          "order": [
            "title",
            "content",
            "section"
          ]
        },
        "values": {
          "content": {
            "sortRemainingBy": "count"
          },
          "title": {
            "sortRemainingBy": "count"
          }
        }
      }
    }
  },
  "rules": [],
  "synonyms": []
}``

### Reproducible demo

https://github.com/ottozastrow/indexAnyFrontend

### Steps to reproduce

npm run build
npm run serve
search something and what console network log

### Expected behavior

display algolia search results

### Actual behavior

displays nothing in search widget and throws error on console.

### Your environment

- Public source code:
- Public site URL:
- Docusaurus version used:
- Environment name and version (e.g. Chrome 89, Node.js 16.4):
- Operating system and version (e.g. Ubuntu 20.04.2 LTS):

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by Josh-Cena** [comment]
We only read `hierarchy` in `SearchPage`:

https://github.com/facebook/docusaurus/blob/1aae05b606238259088410ac729f41ee3cd559d5/packages/docusaurus-theme-search-algolia/src/theme/SearchPage/index.tsx#L240-L242

And there are a multitude of known issues with the search page. Are you getting this on the `/search` page? Or in the search modal as well?

**Comment by ottozastrow** [comment]
getting this outside the /search page aswell

**Comment by slorber** [comment]
I'm running your repro in StackBlitz, but Algolia answers 200 with 0 search hit.

https://stackblitz.com/github/ottozastrow/indexAnyFrontend/tree/main/standlet

What am I supposed to type in the search exactly?

**Comment by ottozastrow** [comment]
after some initial confusion I noticed the remote repo wasn't on the right commit
now you should can reproduce on stackblitz / live site. 
the live site is here: [https://eu-medical-regulations.netlify.app/](https://eu-medical-regulations.netlify.app/)

**Comment by slorber** [comment]
Thanks, I can see the errors indeed.

I don't think this is a Docusaurus bug so I'm going to close for now.

This looks related to your Algolia index (maybe a crawler config problem?) because if you use the Docusauurs v2 index it works, so definitively related to the returned search results.

```
    // Fails
    {
      algolia: {
        appId: 'VAV5CJH9ZY',
        apiKey: 'c427f3cd1975141cf295efa3c2e71ec9',
        indexName: 'mdr_eu745',
        contextualSearch: false,
      },
      // Works
      algolia: {
        appId: 'X1Z85QJPUV',
        apiKey: 'bf7211c161e8205da2f933a02534105a',
        indexName: 'docusaurus-2',
        contextualSearch: false,
      },
```

Maybe try to compare the json of both search results until you notice a difference?

@shortcuts any idea?

**Comment by shortcuts** [comment]
Hey, this indeed seems to be an indexing issue. How is your site indexed? @ottozastrow

**Comment by ottozastrow** [comment]
algolia asks to send them the formatted data.
the format my records have is
```
sample = {
                "title": titles[i],
                "section": chapters[i],
                "content": section,
                "url": f"{page_url}/{chapters[i]}/{document_ids[i]}"
            }
```
(Seems like in the past there was only their crawler, which is now proprietary)

**Comment by ottozastrow** [comment]
here my indexes config.json on algolia
```
{
  "settings": {
    "minWordSizefor1Typo": 4,
    "minWordSizefor2Typos": 8,
    "hitsPerPage": 20,
    "maxValuesPerFacet": 100,
    "searchableAttributes": [
      "unordered(content)",
      "unordered(section)",
      "unordered(title)",
      "unordered(url)"
    ],
    "numericAttributesToIndex": null,
    "attributesToRetrieve": null,
    "allowTyposOnNumericTokens": false,
    "unretrievableAttributes": null,
    "optionalWords": null,
    "queryLanguages": [],
    "attributesForFaceting": [
      "content",
      "section",
      "title"
    ],
    "attributesToSnippet": null,
    "attributesToHighlight": null,
    "paginationLimitedTo": 1000,
    "attributeForDistinct": "permalink",
    "exactOnSingleWordQuery": "attribute",
    "ranking": [
      "typo",
      "geo",
      "words",
      "filters",
      "proximity",
      "attribute",
      "exact",
      "custom"
    ],
    "customRanking": null,
    "separatorsToIndex": "",
    "removeWordsIfNoResults": "none",
    "queryType": "prefixLast",
    "highlightPreTag": "<em>",
    "highlightPostTag": "</em>",
    "snippetEllipsisText": "",
    "alternativesAsExact": [
      "ignorePlurals",
      "singleWordSynonym"
    ],
    "indexLanguages": [],
    "renderingContent": {
      "facetOrdering": {
        "facets": {
          "order": [
            "title",
            "content",
            "section"
          ]
        },
        "values": {
          "content": {
            "sortRemainingBy": "count"
          },
          "title": {
            "sortRemainingBy": "count"
          }
        }
      }
    }
  },
  "rules": [],
  "synonyms": []
}
```

**Comment by shortcuts** [comment]
> algolia asks to send them the formatted data.

Indeed, but DocSearch requires a hierarchical structure of records to reflect it in the UI, we have a small description here: https://docsearch.algolia.com/docs/legacy/how-do-we-build-an-index/

> (Seems like in the past there was only their crawler, which is now proprietary)

The open-source scraper is still available if you'd like to use it or replicate the indexing/record structure. It's hosted on the algolia/docsearch-scraper repository

**Comment by ottozastrow** [comment]
ah interesting - will try that. But I understand correctly that in theory both methods should work (docsearch crawler vs me using my existing structure). But that I should try the other (docsearch), that is kind of "legacy" but still in practice

**Comment by shortcuts** [comment]
> ah interesting - will try that. But I understand correctly that in theory both methods should work (docsearch crawler vs me using my existing structure).

You can use any way to index data, but the record structure sent to Algolia must match the one read by the frontend library (hierarchical). The alternative of keeping your structure would be to apply logic on the frontend (via transformItems for example), but it would have an impact on your performances/users. 

> But that I should try the other (docsearch), that is kind of "legacy" but still in practice

It's not what we use in prod anymore so it's legacy but would still do the job for most of the websites

**Comment by ottozastrow** [comment]
ok thanks!!
